cli/azd/pkg/cmdsubst/containerAppImage.go

```diff
@@ -0,0 +1,110 @@
+//	}
+//
+// Here, we use the value of SERVICE_WEB_IMAGE_NAME if it is set (this value is updated in the .env file every time
+// `azd deploy` is run) and if it is not set (for example, in CI where this value is not persisted across runs), the
+// current value is pulled from the live Container App resource.
+//
+// This can be useful to ensure that the container app image is not reset to an initial value.
+const ContainerAppImageCommandName string = "containerAppImage"
+
+type ContainerAppImageCommandExecutor struct {
+	resourceManager      project.ResourceManager
+	containerAppsService containerapps.ContainerAppService
+	subscriptionId       string
+	services             map[string]*project.ServiceConfig
+}
+
+func NewContainerAppImageExecutor(
+	resourceManager project.ResourceManager,
+	containerAppsService containerapps.ContainerAppService,
+	subscriptionId string,
+	services map[string]*project.ServiceConfig) *ContainerAppImageCommandExecutor {
+	return &ContainerAppImageCommandExecutor{
+		resourceManager:      resourceManager,
+		containerAppsService: containerAppsService,
+		subscriptionId:       subscriptionId,
+		services:             services,
+	}
+}
+
+func (e *ContainerAppImageCommandExecutor) Run(
+	ctx context.Context,
+	commandName string,
+	args []string,
+) (bool, string, error) {
+	if commandName != ContainerAppImageCommandName {
+		return false, "", nil
+	}
+
+	// We expect two arguments, the name of the service to target and the name of the container to pull the image of
+	if len(args) != 2 {
+		return false, "", fmt.Errorf("%s expected one two arguments but got %d", ContainerAppImageCommandName, len(args))
+	}
+
+	serviceName := args[0]
+	containerName := args[1]
+
+	if ctx == nil || e.resourceManager == nil {
+		// Should never happen really...
+		return false, "", fmt.Errorf("missing context information for %s command", SecretOrRandomPasswordCommandName)
+	}
+
+	service, has := e.services[args[0]]
+	if !has {
+		return false, "", fmt.Errorf("%s no service %s defined in azure.yaml", ContainerAppImageCommandName, serviceName)
+	}
+
+	// If the target has not been created yet, return an empty image.
+	targetResource, err := e.resourceManager.GetTargetResource(ctx, e.subscriptionId, service)
+	if err != nil {
+		return true, "", nil
+	}
+
+	if !strings.EqualFold(targetResource.ResourceType(), string(infra.AzureResourceTypeContainerApp)) {
+		return false, "", fmt.Errorf("service %s does not target container apps", serviceName)
+	}
+
+	app, err := e.containerAppsService.GetContainerApp(
+		ctx,
+		targetResource.SubscriptionId(),
+		targetResource.ResourceGroupName(),
+		targetResource.ResourceName(),
+	)
+	if err != nil {
+		return false, "", fmt.Errorf("reading container app properties: %w", err)
+	}
+
+	for _, container := range app.Properties.Template.Containers {
+		if *container.Name == containerName {
+			return true, *container.Image, nil
+		}
+	}
```

Comment:
I have to put forward this idea to see where we land on this.

I think it's entirely possible to replicate this exact same logic inside `bicep`. Supposed if they were equivalent; I would much prefer that instead of adding additional complexity and features into the CLI.

Now I've been hearing some concerns about using `bicep` this way, but I have not found concrete evidence that raises concerns. Behind the scenes, the ARM deployment service is going to perform the same exact REST calls we're doing there, except with a smaller critical window where it can immediately issue a GET followed by a subsequent PUT if needed. Unlike the current model, where the observation window is minutes long between the ARM deployment starting and the ARM deployment actually performing the PUT on the resource.

Reply:
I am assuming you are referring to querying the resource with an `existing` keyword to get the container app and to optionally get the existing container image?  This would be great except that on first provision if the resource is not found the ARM deployment will fail with no way to recover.

My original implementation did this but it needed some signal to determine whether or not the resource exists.

Reply:
Wei's initial thought here was that we might be able to use something from the [`list*`](https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/bicep-functions-resource#list]) family of bicep functions, instead of `existing` and do so in a way that did not fail when the resource didn't exist.

But, after we looked into that more together, we both think it is not possible.  We think if we do something in bicep, we'd have to take the approach you outlined, and pass some information into bicep so we can decide if we want to do the `existing` read or not, since it would fail if the resource didn't exist.

Reply:
@weikanglim Just wanted to confirm that what I wrote above matches your understanding here.  I think we either need to do something like this or do something like what Wallace outlined where we set a value to know if it's safe to use `existing` or not in our bicep.